parabolic: 2024.5.0 -> 2025.1.4

normalcea opened the pull request:
Updates parabolic's package derivation to support the new C++ code base and dependencies (libnick, libxmlxx5 and libmaddy-markdown).

Parabolic supports both a Qt and GNOME interface (qt6 and libadwaita respectively). Choosing the Qt interface requires overriding the package derivation.

This a redo of PR #363469 which has grown stale (I've also learned a ton of Nix-fu since then)

I will be taking maintainership of the parabolic package from now on as @EwuUwe has not had any activity since the past year going by git history. If @getchoo wishes to be part of the maintainers list then I'll add them in as they did last time.

### Caveats
The Qt interface has a documentation button that is supposed to show KDE documentation for the program. I can't seem to get it working even after adding `kdePackages.kdoctool` to `nativeBuildInputs`.  This issue can be replicated by using the program in KDE Plasma and clicking the documentation tooltip button.

The Qt interface can only be built if they user overrides the package derivation, so I added a note in `longDescription` for that. 

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Commit:
libmaddy-markdown: init at 1.3.0

Commit:
libxmlxx5: init at 5.4

Commit:
libnick: init at 2025.1.0

Commit:
parabolic: 2024.5.0 -> 2025.1.4

SuperSandro2000 commented on pkgs/by-name/li/libmaddy-markdown/package.nix:
```diff
@@ -0,0 +1,40 @@
+    description = "C++ Markdown to HTML header-only parser library";
+    homepage = "https://github.com/progsource/maddy";
+    license = lib.licenses.mit;
+    maintainers = [ ];
```
This must be set

SuperSandro2000 commented on pkgs/by-name/li/libnick/package.nix:
```diff
@@ -0,0 +1,86 @@
+  src = fetchFromGitHub {
+    owner = "NickvisionApps";
+    repo = "libnick";
+    tag = "${finalAttrs.version}";
```
```suggestion
    tag = finalAttrs.version;
```

SuperSandro2000 commented on pkgs/by-name/li/libnick/package.nix:
```diff
@@ -0,0 +1,86 @@
+  nix-update-script,
+  withLibsecret ? stdenv.hostPlatform.isUnix && !stdenv.hostPlatform.isDarwin,
+}:
+assert lib.assertMsg (
+  (!withLibsecret) -> stdenv.hostPlatform.isDarwin
+) "libsecret is only optional on Darwin";
```
```suggestion
```
I would say we don't expose it if we need an assertion

SuperSandro2000 commented on pkgs/by-name/li/libxmlxx5/package.nix:
```diff
@@ -0,0 +1,40 @@
+  doCheck = true;
+
+  meta = {
+    description = "C++ wrapper for the libxml2 XML parser library, version 5";
```
```suggestion
    description = "C++ wrapper for the libxml2 XML parser library";
```

SuperSandro2000 commented on pkgs/by-name/li/libxmlxx5/package.nix:
```diff
@@ -0,0 +1,40 @@
+    description = "C++ wrapper for the libxml2 XML parser library, version 5";
+    homepage = "https://libxmlplusplus.sourceforge.net/";
+    license = lib.licenses.lgpl2Plus;
+    maintainers = [ ];
```
A maintainer must be set

SuperSandro2000 commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,125 @@
-    rev = version;
-    hash = "sha256-awbCn7W7RUSuEByXxLGrsmYjmxCrwywhhrMJq/iM1Uc=";
-    fetchSubmodules = true;
+    tag = "${finalAttrs.version}";
```
```suggestion
    tag = finalAttrs.version;
```

SuperSandro2000 commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,125 @@
+      pkg-config
+    ]
+    ++ (
+      if !useQtInterface then
```
Lets invert the condition to not be inverted

SuperSandro2000 commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,125 @@
+      boost
+    ]
+    ++ (
+      if !useQtInterface then
```
same here

SuperSandro2000 commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,125 @@
-    gdk-pixbuf
-    blueprint-compiler
+  cmakeFlags = [
+    (lib.cmakeFeature "UI_PLATFORM" (if !useQtInterface then "gnome" else "qt"))
```
and here

SuperSandro2000 commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,125 @@
 
-  buildInputs = [ gtk4 libadwaita ];
+  preFixup =
+    (if !useQtInterface then "gappsWrapperArgs" else "qtWrapperArgs")
```
and here

SuperSandro2000 commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,125 @@
+  preFixup =
+    (if !useQtInterface then "gappsWrapperArgs" else "qtWrapperArgs")
+    + ''
+      +=(
+            --prefix PATH : ${
+              lib.makeBinPath [
+                aria2
+                ffmpeg
+                yt-dlp
+              ]
+            }
+          )
+    '';
```
```suggestion
      +=(--prefix PATH : ${lib.makeBinPath [ aria2 ffmpeg yt-dlp ]})
    '';
```

normalcea commented:
Implemented all recommendations and added myself to the maintainers list of above packages. 

Thank you for the code review.

getchoo commented on pkgs/by-name/li/libnick/package.nix:
```diff
@@ -0,0 +1,81 @@
+
+  postPatch = ''
+    substituteInPlace cmake/libnick.pc.in \
+    --replace-fail 'libdir=''${exec_prefix}/@CMAKE_INSTALL_LIBDIR@' 'libdir=@CMAKE_INSTALL_FULL_LIBDIR' \
+    --replace-fail 'includedir=''${prefix}/@CMAKE_INSTALL_INCLUDEDIR@' 'includedir=@CMAKE_INSTALL_FULL_INCLUDEDIR'
```
```suggestion
    --replace-fail 'libdir=''${exec_prefix}/@CMAKE_INSTALL_LIBDIR@' 'libdir=@CMAKE_INSTALL_FULL_LIBDIR@' \
    --replace-fail 'includedir=''${prefix}/@CMAKE_INSTALL_INCLUDEDIR@' 'includedir=@CMAKE_INSTALL_FULL_INCLUDEDIR@'
```
I'm *pretty sure* these are needed

getchoo commented on pkgs/by-name/li/libxmlxx5/package.nix:
```diff
@@ -0,0 +1,40 @@
+
+  propagatedBuildInputs = [ libxml2 ];
+
+  doCheck = true;
```
```suggestion
  doCheck = stdenv.buildPlatform.canExecute stdenv.hostPlatform;
```
https://github.com/NixOS/nixpkgs/blob/c3b4161dac4a30a098a5c35767b1f0b41916d52f/pkgs/stdenv/generic/make-derivation.nix#L286-L288

getchoo commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,119 @@
+  ffmpeg,
+  aria2,
+  nix-update-script,
+  useQtInterface ? false,
```
```suggestion
  uiPlatform ? "gnome",
```
I think it would be better to make this an enum, similar to how it's modeled in CMake. This will make the expression a bit less verbose, and more flexible in case any other UI platforms get added

normalcea replied:
I agree, having this as an enum is more elegant of a solution.

getchoo commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,119 @@
+  nix-update-script,
+  useQtInterface ? false,
 }:
 
```
```suggestion

assert lib.assertOneOf "uiPlatform" uiPlatform [ "gnome" "qt" ];

```
We can also type check it

getchoo commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,119 @@
+      ninja
+      pkg-config
+    ]
+    ++ (
+      if !useQtInterface then
```
```suggestion
    ++ lib.optionals (uiPlatform == "gnome")
```

getchoo commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,119 @@
+          glib
+          shared-mime-info
+        ]
+      else
+        [ qt6.wrapQtAppsHook ]
```
```suggestion
      ++ lib.optional (uiPlatform == "qt") qt6.wrapQtAppsHook;
```

getchoo commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,119 @@
+      libnick
+      boost
+    ]
+    ++ (
+      if useQtInterface then
```
```suggestion
    ++ lib.optionals (uiPlatform == "qt")
```

getchoo commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,119 @@
+          qt6.qtbase
+          qt6.qtsvg
+        ]
+      else
+        [
```
```suggestion
      ++ lib.optionals (uiPlatform == "gnome") [
```

getchoo commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,119 @@
-    gdk-pixbuf
-    blueprint-compiler
+  cmakeFlags = [
+    (lib.cmakeFeature "UI_PLATFORM" (if useQtInterface then "qt" else "gnome"))
```
```suggestion
    (lib.cmakeFeature "UI_PLATFORM" uiPlatform)
```

getchoo commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,119 @@
   ];
 
-  buildInputs = [ gtk4 libadwaita ];
+  preFixup =
+    (if useQtInterface then "qtWrapperArgs" else "gappsWrapperArgs")
+    + ''+=(--prefix PATH : ${
+      lib.makeBinPath [
+        aria2
+        ffmpeg
+        yt-dlp
+      ]
+    })'';
```
```suggestion
  preFixup =
    lib.optionalString (uiPlatform == "gnome") "gappsWrapperArgs"
    + lib.optionalString (uiPlatform == "qt") "qtWrapperArgs"
    + "+=(--prefix PATH : ${
      lib.makeBinPath [
        aria2
        ffmpeg
        yt-dlp
      ]
    })";
```

getchoo commented on pkgs/by-name/pa/parabolic/package.nix:
```diff
@@ -1,77 +1,119 @@
-    mainProgram = "NickvisionTubeConverter.GNOME";
-    platforms = platforms.linux;
+    license = lib.licenses.gpl3Plus;
+    maintainers = [ lib.maintainers.normalcea ];
```
```suggestion
    maintainers = with lib.maintainers; [
      normalcea
      getchoo
    ];
```

Would still love to help out :)

getchoo reviewed (commented):
Sorry the previous PR got lost in my inbox. This is looking great!

normalcea commented:
Implemented all recommendations (as well as updating parabolic's `longDescription` to match) .

I also took the liberty of bumping the release to 2024.1.4.

SuperSandro2000 commented on pkgs/by-name/li/libxmlxx5/package.nix:
```diff
@@ -0,0 +1,40 @@
+
+  src = fetchurl {
+    url = "https://download.gnome.org/sources/libxml++/${version}/libxml++-${lib.versions.pad 3 version}.tar.xz";
+    sha256 = "sha256-6aI8Q2aGqUaY0hOOa8uvhJEh1jv6D1DcNP77/XlWaEg=";
```
```suggestion
    hash = "sha256-6aI8Q2aGqUaY0hOOa8uvhJEh1jv6D1DcNP77/XlWaEg=";
```

normalcea commented:
libxmlxx5 fixed!

normalcea commented:
Moved `itstool`  and  `yelp-tools` outside of gnome optional since they are referenced and used in the root `CMakeLists.txt` of the project for both qt and gnome. 

```diff
modified   pkgs/by-name/pa/parabolic/package.nix
@@ -46,10 +46,10 @@ stdenv.mkDerivation (finalAttrs: {
       gettext
       ninja
       pkg-config
-    ]
-    ++ lib.optionals (uiPlatform == "gnome") [
       itstool
       yelp-tools
+    ]
+    ++ lib.optionals (uiPlatform == "gnome") [
       wrapGAppsHook4
       blueprint-compiler
       glib
```

SuperSandro2000 committed:
parabolic: cleanup string interpolation
